lib/gollum-lib/wiki.rb

```diff
@@ -504,31 +504,14 @@ def size(ref = nil)
     # query - The string to search for
     #
     # Returns an Array with Objects of page name and count of matches
-    def search(query)
+    def search(query, return_terms = false)
       options = {:path => page_file_dir, :ref => ref}
-      results = {}
-      @repo.git.grep(query, options).each do |hit|
-        name = hit[:name]
-        count = hit[:count]
-        # Remove ext only from known extensions.
-        # test.pdf => test.pdf, test.md => test
-        file_name = Page::valid_page_name?(name) ? name.chomp(::File.extname(name)) : name
-        results[file_name] = count.to_i
-      end
-
-      # Use ls_files '*query*' to search for file names. Grep only searches file content.
-      # Spaces are converted to dashes when saving pages to disk.
-      @repo.git.ls_files(query, options).each do |path|
-        # Remove ext only from known extensions.
-        file_name          = Page::valid_page_name?(path) ? path.chomp(::File.extname(path)) : path
-        # If there's not already a result for file_name then
-        # the value is nil and nil.to_i is 0.
-        results[file_name] = results[file_name].to_i + 1;
-      end
-
-      results.map do |key, val|
-        { :count => val, :name => key }
+      search_terms = query.scan(/"([^"]+)"|(\S+)/).flatten.compact.map {|term| Regexp.escape(term)}
+      results = @repo.git.grep(search_terms, options).map do |result|
+        result[:name] = extract_page_file_dir(result[:name])
+        result
       end
+      return_terms ? [results, search_terms] : results
```

Comment:
I would prefer it if the returned object is always of the same data structure, so I would have the else clause return `[results, nil]` or `[results, []]` and always assume that search returns a two-placed array.